### PR TITLE
Allow scene nodes to have multiple scripts

### PR DESCRIPTION
**Pull request by dasimonde**
Nodes can have multiple scripts now

### CODE REVIEW

**Comment by dasimonde** [comment]
`error[E0499]: cannot borrow `context` as mutable more than once at a time
   --> src/engine/mod.rs:908:50
    |
903 |                     if let Some(node) = context.scene.graph.try_get_mut(handle_node) {
    |                                         ------------------- first mutable borrow occurs here
...
908 |                                 script.on_deinit(&mut context)
    |                                                  ^^^^^^^^^^^^ second mutable borrow occurs here
...
912 |                         for script in node
    |                                       ---- first borrow later used here

error[E0499]: cannot borrow `context` as mutable more than once at a time
   --> src/engine/mod.rs:921:50
    |
903 |                       if let Some(node) = context.scene.graph.try_get_mut(handle_node) {
    |                                           ------------------- first mutable borrow occurs here
...
912 |                           for script in node
    |  _______________________________________-
913 | |                             .secondary_scripts
914 | |                             .iter_mut()
915 | |                             .map(|i| i.0.as_mut())
916 | |                             .flatten()
    | |______________________________________- first borrow later used here
...
921 |                                   script.on_deinit(&mut context);
    |                                                    ^^^^^^^^^^^^ second mutable borrow occurs here

error[E0499]: cannot borrow `*context` as mutable more than once at a time
    --> src/engine/mod.rs:1067:30
     |
1052 |             let node = match context.scene.graph.try_get_mut(context.handle) {
     |                              ------------------- first mutable borrow occurs here
...
1067 |                 func(script, context);
     |                              ^^^^^^^ second mutable borrow occurs here
1068 |             };
1069 |             node.script = script;
     |             ---- first borrow later used here
...
1088 | define_process_node!(process_node, ScriptContext);
     | ------------------------------------------------- in this macro invocation
     |
     = note: this error originates in the macro `define_process_node` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0499]: cannot borrow `*context` as mutable more than once at a time
    --> src/engine/mod.rs:1078:35
     |
1052 |             let node = match context.scene.graph.try_get_mut(context.handle) {
     |                              ------------------- first mutable borrow occurs here
...
1071 |             for vec_script in node.secondary_scripts.iter_mut() {
     |                               --------------------------------- first borrow later used here
...
1078 |                 func(&mut script, context);
     |                                   ^^^^^^^ second mutable borrow occurs here
...
1088 | define_process_node!(process_node, ScriptContext);
     | ------------------------------------------------- in this macro invocation
     |
     = note: this error originates in the macro `define_process_node` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0499]: cannot borrow `*context` as mutable more than once at a time
    --> src/engine/mod.rs:1067:30
     |
1052 |             let node = match context.scene.graph.try_get_mut(context.handle) {
     |                              ------------------- first mutable borrow occurs here
...
1067 |                 func(script, context);
     |                              ^^^^^^^ second mutable borrow occurs here
1068 |             };
1069 |             node.script = script;
     |             ---- first borrow later used here
...
1089 | define_process_node!(process_node_message, ScriptMessageContext);
     | ---------------------------------------------------------------- in this macro invocation
     |
     = note: this error originates in the macro `define_process_node` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0499]: cannot borrow `*context` as mutable more than once at a time
    --> src/engine/mod.rs:1078:35
     |
1052 |             let node = match context.scene.graph.try_get_mut(context.handle) {
     |                              ------------------- first mutable borrow occurs here
...
1071 |             for vec_script in node.secondary_scripts.iter_mut() {
     |                               --------------------------------- first borrow later used here
...
1078 |                 func(&mut script, context);
     |                                   ^^^^^^^ second mutable borrow occurs here
...
1089 | define_process_node!(process_node_message, ScriptMessageContext);
     | ---------------------------------------------------------------- in this macro invocation
     |
     = note: this error originates in the macro `define_process_node` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0499]: cannot borrow `*scene` as mutable more than once at a time
    --> src/engine/mod.rs:1928:41
     |
1916 |                         if let Some(node) = scene.graph.try_get_mut(node_task_handler.node_handle) {
     |                                             ----------- first mutable borrow occurs here
...
1928 |                                         scene,
     |                                         ^^^^^ second mutable borrow occurs here
...
1940 |                                 node.script.replace(script);
     |                                 ---- first borrow later used here

error[E0499]: cannot borrow `*scene` as mutable more than once at a time
    --> src/engine/mod.rs:1953:45
     |
1916 |                         if let Some(node) = scene.graph.try_get_mut(node_task_handler.node_handle) {
     |                                             ----------- first mutable borrow occurs here
...
1943 |                             for vec_script in node.secondary_scripts.iter_mut() {
     |                                               --------------------------------- first borrow later used here
...
1953 |                                             scene,
     |                                             ^^^^^ second mutable borrow occurs here
`

**Comment by dasimonde** [comment]
Everything should work now. The scripts for nodes in old scenes are loaded correctly now. Also I removed the Option type for the script buffer and combine script variable and the script buffer variable in one variable.

**Comment by mrDIMAS** [comment]
Could you please also fix tests compilation and ensure that they pass? Also CI fails on documentation, run `cargo doc --all-features` and fix the warnings/errors too. This PR changes fundamental part of the engine and it must be perfect to not break existing projects.

File: src/scene/base.rs [hunk 1]
```diff
@@ -727,113 +752,184 @@ impl Base {
...
+    pub fn add_script(&mut self, script: Script) {
+        self.scripts.push(ScriptWrapper(Some(script)));
+        if let Some(sender) = self.script_message_sender.as_ref() {
+            Log::verify(sender.send(NodeScriptMessage::InitializeScript {
+                handle: self.self_handle,
+            }));
+        }
+    }
+
+    fn set_scripts_internal(&mut self, (index, script): (usize, Option<Script>)) -> Option<Script> {
```
Comment (mrDIMAS) [inline src/scene/base.rs:824]: This method is used in reflection-based access to `scripts` field and its signature must be `fn set_scripts_internal(&mut self, scripts: Vec<ScriptWrapper>) -> Vec<ScriptWrapper>`

File: src/scene/base.rs [hunk 2]
```diff
@@ -973,11 +1073,58 @@ impl Visit for Base {
         //
         // None of the reasons are fatal and we should still give an ability to load such node
         // to edit or remove it.
-        if let Err(e) = visit_opt_script("Script", &mut self.script, &mut region) {
-            // Do not spam with error messages if there is missing `Script` field. It is ok
-            // for old scenes not to have script at all.
-            if !matches!(e, VisitError::RegionDoesNotExist(_)) {
-                Log::err(format!("Unable to visit script. Reason: {:?}", e))
+
+        if region.is_reading() {
```
Comment (mrDIMAS) [inline src/scene/base.rs:1077]: These two branches of `if` are way too convoluted. To preserve backward compatibility all you need to do is:

```rust
let old_script: Option<Script> = None;
if region.is_reading() && old_script.visit("Script", region).is_ok() {
  self.scripts.push(ScriptWrapper(old_script));
  return;
}

self.scripts.visit("Scripts", region)?;
```

File: src/scene/graph/mod.rs [hunk 3]
```diff
@@ -438,15 +438,20 @@ impl Graph {
         self.link_nodes(child, parent);
     }
 
-    /// Searches for a **first** node with a script of the given type `S` in the hierarchy starting from the
+    /// Searches for a **first** node with a secondary script of the given type `S` in the hierarchy starting from the
```
Comment (mrDIMAS) [inline src/scene/graph/mod.rs:441]: Remove `secondary`

File: src/scene/graph/mod.rs [hunk 4]
```diff
@@ -1353,43 +1358,52 @@ impl Graph {
...
         self.try_get_mut(node)
-            .and_then(|node| node.try_get_script_mut())
+            .and_then(|node| node.try_get_script_mut(index))
     }
 
     /// Tries to borrow a node using the given handle and fetch a reference to a component of the given type
     /// from the script of the node.
     #[inline]
-    pub fn try_get_script_component_of<C>(&self, node: Handle<Node>) -> Option<&C>
+    pub fn try_get_secondary_script_component_of<C>(
```
Comment (mrDIMAS) [inline src/scene/graph/mod.rs:1382]: Revert the name back.

**Review by mrDIMAS — CHANGES_REQUESTED** [review]

**Comment by dasimonde** [comment]
I have implemented all your change requests and all CI tests are green now

File: src/scene/graph/mod.rs [hunk 5]
```diff
@@ -1353,43 +1358,48 @@ impl Graph {
 
     /// Tries to borrow a node using the given handle, fetch its script and cast it to the specified type.
     #[inline]
-    pub fn try_get_script_of<T>(&self, node: Handle<Node>) -> Option<&T>
+    pub fn try_get_script_of<T>(&self, node: Handle<Node>, index: usize) -> Option<&T>
```
Comment (mrDIMAS) [inline src/scene/graph/mod.rs:1361]: This method and every other that accepts `T` as a type argument, should search for a first script of type `T`. Similar to Unity's `GetComponent` (https://docs.unity3d.com/ScriptReference/GameObject.GetComponent.html) . Also the doc comment above the method should clearly state that this and other methods search for the first script of the given type.

**Comment by dasimonde** [comment]
I hope everything is fine now

**Review by mrDIMAS — APPROVED** [review]

**Pull request merged**